p2p/src/main/kotlin/maru/p2p/discovery/DiscoveryService.kt

```diff
@@ -0,0 +1,187 @@
+        .map { NodeRecordFactory.DEFAULT.fromEnr(it) }
+        .toList()
+
+    discoveryNetworkBuilder.listen(p2pConfig.ipAddress, p2pConfig.discoveryPort.toInt())
+    discoveryNetworkBuilder.secretKey(privateKey)
+    discoveryNetworkBuilder.localNodeRecord(localNodeRecord)
+    discoveryNetworkBuilder.bootnodes(bootnodes)
+
+    discoverySystem = discoveryNetworkBuilder.build()
+
+    val maruForkId = MaruForkId.MARU_INITIAL_FORK_ID // TODO: set the correct fork id here
+
+    discoverySystem
+      .start()
+      .thenRun {
+        discoverySystem.updateCustomFieldValue(MaruForkId.MARU_FORK_ID_FIELD_NAME, maruForkId.encode())
+        // TODO: do we want another custom field to identify topics/role/something else?
+        this.bootnodeRefreshTask =
+          delayedExecutor.runWithFixedDelay(
+            { this.pingBootnodes(bootnodes) },
+            BOOTNODE_REFRESH_DELAY,
+            { error: Throwable? ->
+              log.error(
+                "Failed to contact discovery bootnodes",
+                error,
+              )
+            },
+          )
+      }
+  }
+
+  fun stop() {
+    discoverySystem.stop()
+  }
+
+  fun searchForPeers(): SafeFuture<Collection<MaruDiscoveryPeer>> =
+    SafeFuture
+      .of(discoverySystem.searchForNewPeers())
+      // Current version of discovery doesn't return the found peers but next version will
+      .thenApply { getKnownPeers() }
+
+  fun getKnownPeers(): Collection<MaruDiscoveryPeer> =
+    discoverySystem
+      .streamLiveNodes()
+      .map { node: NodeRecord ->
+        convertNodeRecordToDiscoveryPeer(node)
+      }.toList()
+
+  fun getLocalNodeRecord(): NodeRecord = localNodeRecord
+
+  private fun convertNodeRecordToDiscoveryPeer(node: NodeRecord): MaruDiscoveryPeer =
+    MaruDiscoveryPeer(
+      (node.get(EnrField.PKEY_SECP256K1) as Bytes),
+      node.nodeId,
+      node.tcpAddress.get(),
+      MaruForkId.fromNodeRecord(node),
+    )
+
+  private fun pingBootnodes(bootnodeRecords: List<NodeRecord>) {
+    bootnodeRecords.forEach(
+      Consumer { bootnode: NodeRecord? ->
+        SafeFuture
+          .of(discoverySystem.ping(bootnode))
+          .exceptionally {
+            log.info("Bootnode {} is unresponsive", bootnode)
+            println("Bootnode ${bootnode?.nodeId} is unresponsive")
+            throw it
+          }
+      },
+    )
+  }
+
+  private fun localNodeRecord(privateKey: SecretKey): NodeRecord {
+    val nodeRecordBuilder: NodeRecordBuilder =
+      NodeRecordBuilder()
+        .secretKey(privateKey)
+        .seq(UInt64.ONE)
+        .address(
+          p2pConfig.ipAddress,
+          p2pConfig.discoveryPort.toInt(),
+        ).customField(MaruForkId.MARU_FORK_ID_FIELD_NAME, MaruForkId.MARU_INITIAL_FORK_ID.encode())
+    return nodeRecordBuilder.build()
+  }
+
+  data class MaruDiscoveryPeer(
+    val publicKey: Bytes,
+    val nodeId: Bytes,
+    val address: InetSocketAddress,
+    val maruForkId: MaruForkId,
+  )
+
+  class MaruForkId(
```

Comment:
1. This looks like a data object
2. I think it has to be defined somewhere in the core and be used in P2P communication as well as a part of topic id

Reply:
Yes! I will use the right one now :-)